fix to ao bitmap scan bug

bucklerjj opened the pull request:
It looks like there is a bug in the append-optimized columnar bitmap scan logic when using a b-tree index.  The issue is that a bitmap scan can become lossy if the row-level bitmap becomes too large and is partially promoted to a page-level bitmap.  In that case, the current scan logic does not necessarily have all the columns to recheck the filter.  This occurs when the query filters on an indexed column, but does not select it.  The issue does not affect heap or append-optimized row oriented tables since all of the columns are available to reevaluate.  I am not sure if there are any query plans where a physical bitmap index could be converted to an in-memory bitmap index and suffer the same issue.

gcaragea commented:
Can you also add a ICG test for this? We shouldn't commit code without tests. 
Thanks.

asimrp commented:
Yes, adding ICG is important.  We may have the same test case that you've identified in the mail thread as ICG.

ashwinstar commented:
@gcaragea or @foyzur Please can you make sure to commit this to 4.3_STABLE as well when pushing to Master. Thank You.

bucklerjj commented:
I added an ICG test called bitmapscan_ao.  Give me a heads-up if the test does not follow guidelines/norms.

asimrp commented:
@bucklerjj, this fix is adding more columns to be scanned from storage even when they need not be returned to the client.  Can this overhead be minimized by including these columns (`node->bitmapqualorig`) only when we have a lossy tid bitmap.  If the bitmap is not lossy, then the columns need not be scanned.  What do you think?

asimrp commented on src/test/regress/sql/bitmapscan_ao.sql:
```diff
@@ -0,0 +1,26 @@
+drop table if exists bms_ao_bug;
```
Thank you for the test.  You have followed the norm correctly.  Just a couple of requests:
1. I think a comment here about intension of the test, which is to verify column oriented scan when tid bitmap is lossy.
2. Can we move this test to `test/regress/input/aocs.source`?  That's the place for basic column oriented tests.

We use "aocs" suffix to name test files specific for column oriented tables.  E.g. `

bucklerjj commented:
I put out a version of the bitmap scan that creates two fetch descriptors.  One for the lossy case and one for the non-lossy case.  The BitmapAppendOnlyScanNext function then alternates between them as necessary.  It seems about 15% faster for the two-column case where one of the two columns does not need to be fetched but is filtered on.  Thoughts?

Another approach might be to scrap and recreate the fetch descriptor as lossy if and when a lossy portion is encountered.  A third approach would be to lazily create the lossy descriptor when needed.  Either would complicate the code a bit more, but might save a bit of initialization time.

I also updated the aocs.source ICG test and removed the bitmapscan_ao test.

asimrp commented:
Thank you for the two descriptor version and also for changing the test.  Hmm, 50% less I/O results in 15% performance gain.  Does not appear very promising.  But at least something is better than nothing.  Without losing too much of simplicity, I think this version can be committed.

:+1: from my side.

bucklerjj commented:
Just a clarification.  The two columns were both constant integers that
were compressed. So there was no real IO.  All the savings came from a
little less cpu work to populate the slot.

Jay Buckler
On Feb 19, 2016 8:38 PM, "Asim RP" <notifications@github.com> wrote:

> Thank you for the two descriptor version and also for changing the test.
> Hmm, 50% less I/O results in 15% performance gain. Does not appear very
> promising. But at least something is better than nothing. Without losing
> too much of simplicity, I think this version can be committed.
>
> [image: :+1:] from my side.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/376#issuecomment-186478564>.
>

armenatzoglou commented:
@ashwinstar I will push it to 4.3_STABLE after it is merged to master. thanks

foyzur commented on src/backend/executor/nodeBitmapAppendOnlyscan.c:
```diff
@@ -92,9 +94,14 @@ initFetchDesc(BitmapAppendOnlyScanState *scanstate)
 			Assert(currentRelation->rd_att != NULL);
 		
 			proj = (bool *)palloc0(sizeof(bool) * currentRelation->rd_att->natts);
```
May be we can refactor out the size in a separate variable, at least for readability, if not more reliability and maintainability.

foyzur replied:
Please note that nodeBitmapAppendOnlyscan is not used by gp orca optimizer. The optimizer generates a "super" operator BitmapTableScan that encapsulates the underlying storage type. We need to port this fix to execBitmapAOScan.c, e.g.

foyzur commented:
Please note that nodeBitmapAppendOnlyscan is not used by gp orca optimizer. The optimizer generates a "super" operator BitmapTableScan that encapsulates the underlying storage type. This is very useful if you have a heterogenous list of partitions (mixture of AO, AOCO and Heap). We need to port this fix to execBitmapAOScan.c, e.g.

foyzur commented on src/backend/executor/nodeBitmapAppendOnlyscan.c:
```diff
@@ -112,6 +119,24 @@ initFetchDesc(BitmapAppendOnlyScanState *scanstate)
+				projLossy[0] = true;
+			
+			scanstate->baos_currentAOCSLossyFetchDesc =
+				aocs_fetch_init(currentRelation, estate->es_snapshot, appendOnlyMetaDataSnapshot, projLossy);
```
Aren't we leaking memory here? The aocs_fetch_init stores the "proj" into it's member, which I am assuming it frees later. If we are only storing either/or of the proj/projLossy, then we might be leaking.

foyzur commented on src/backend/executor/nodeBitmapAppendOnlyscan.c:
```diff
@@ -112,6 +119,24 @@ initFetchDesc(BitmapAppendOnlyScanState *scanstate)
+			if(colno == currentRelation->rd_att->natts)
+				projLossy[0] = true;
+			
+			scanstate->baos_currentAOCSLossyFetchDesc =
```
I would prefer a refactoring to isolate the logic of iterating proj, and projLossy, and then populating scanstate->baos_currentAOCSLossyFetchDesc and scanState->baos_currentAOCSFetchDesc as the code is identical, except one is working on proj and baos_currentAOCSFetchDesc and the other is working on projLossy and baos_currentAOCSLossyFetchDesc

foyzur commented on src/backend/executor/nodeBitmapAppendOnlyscan.c:
```diff
@@ -138,6 +163,15 @@ freeFetchDesc(BitmapAppendOnlyScanState *scanstate)
+
+	if (scanstate->baos_currentAOCSLossyFetchDesc != NULL)
+	{
+		Assert(!((BitmapAppendOnlyScan *)(scanstate->ss.ps.plan))->isAORow);
```
Identical assertion. And the if block looks very similar. Refactoring?

foyzur commented on src/backend/executor/nodeBitmapAppendOnlyscan.c:
```diff
@@ -138,6 +163,15 @@ freeFetchDesc(BitmapAppendOnlyScanState *scanstate)
+	if (scanstate->baos_currentAOCSLossyFetchDesc != NULL)
+	{
+		Assert(!((BitmapAppendOnlyScan *)(scanstate->ss.ps.plan))->isAORow);
+		pfree(scanstate->baos_currentAOCSLossyFetchDesc->proj);
```
Are you sure you are not intending projLossy? A refactoring perhaps could take care of all.

foyzur commented:
@bucklerjj great job fixing the bug!! I have only two concerns about the fix:

1. The fix should also be put into the execBitmapAOScan (i.e., the new BitmapTableScan operator)
2. Couple of refactorings for future maintainability.

bucklerjj commented:
I don't think the code is leaking memory.  The lossy fetch descriptor is
being created, used, and destroyed in an identical manner to the original
descriptor.

The main reason I created the baos_currentAOCSFetchDesc
and baos_currentAOCSLossyFetchDesc together was to avoid the overhead of
calling the GetNeededColumnsForScan additional times.  If we don't think
that is a big deal, we could move the init logic to another function with a
lossy parameter.

I am not quite clear on the value of refactoring the freeFetchDesc.  The
logic makes changes not only to the descriptor, but also to the scanstate
structure. It also performs an assertion on the scan state.  So we would
still have to pass the scanstate and I guess the address of the pointer
that we wanted to set to null.  It all seems a bit awkward to save 5
admittedly repetitious lines.  We could merge them into the same if clause
(and eliminate the second assertion) since they should always be paired.

The refactoring that has the greatest attraction would be to change the the
AOCS fetch descriptor itself to create optional/lossy columns and then add
a lossy parameter to aocs_fetch.  That would simplify the higher layer of
code and reduce the overhead of maintaining two descriptors with
independent state.  Thoughts?

Jay Buckler

On Tue, Feb 23, 2016 at 6:24 PM, foyzur <notifications@github.com> wrote:

> @bucklerjj <https://github.com/bucklerjj> great job fixing the bug!! I
> have only two concerns about the fix:
>
>    1. The fix should also be put into the execBitmapAOScan (i.e., the new
>    BitmapTableScan operator)
>    2. Couple of refactorings for future maintainability.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/376#issuecomment-187962486>.
>

foyzur commented:
Hi @bucklerjj, I still believe it would be nicer as we always try to refactor similar code, even if they are few lines. That just ensures future maintainability, particularly if we need to fix a bug (or add additional code) and a newcomer does not have the whole picture, it's easier for him/her to overlook all the places. As you mentioned, the true refactoring is in the AOCS fetch descriptor. However, that is a pretty invasive one and I think we should defer that. Anyways, I am fine if you push the code as is. I would suggest you at least raise a issue to investigate if the same problem should be fixed in BitmapTableScan operator.

Thanks,

Foyzur

foyzur commented:
Working on merging the code into master. Lot's of conflicts because of simultaneous changes in test files from two other person. I will resolve the conflicts, ensure the tests are green and then merge.

foyzur commented:
Pushed the changes to master. Closing PR.
